### PR TITLE
Draft PR for rewriting of remaining EKFSym functions in C++

### PR DESCRIPTION
**Pull request by wolterhv**
Created according to https://github.com/commaai/openpilot/issues/2386#issuecomment-815741292

This is a draft PR to provide visibility into my WIP, to avoid rework.

### CODE REVIEW

File: rednose/helpers/ekf_sym.cc [hunk 1]
```diff
@@ -205,3 +205,180 @@ VectorXd EKFSym::update(int kind, VectorXd z, MatrixXdr R, std::vector<double> e
...
+        // matrix
+        to_mult(i, i) = 1;
+        // to_mult[-d4:, :d4] = np.eye(d4)
+        // The submatrix comprised by the last d4 rows and therein all columns up
+        // to the d4-th column is set to an identity matrix
+        if (i < d4) to_mult(red_side+i, i) = 1;
+    }
+
+    // TODO aren't we certain enough that the side of P_reduced is red_side?
+    // Are these asserts really required?
```
Comment (wolterhv) [inline rednose/helpers/ekf_sym.cc:274]: @pd0wm What do you think about removing these asserts, inherited from the python code? See line 248.

File: rednose/helpers/ekf_sym.cc [hunk 2]
```diff
@@ -205,3 +205,180 @@ VectorXd EKFSym::update(int kind, VectorXd z, MatrixXdr R, std::vector<double> e
...
+    assert(P_reduced.rows() == red_side);
+    assert(P_reduced.cols() == red_side);
```
Comment (wolterhv) [inline rednose/helpers/ekf_sym.cc:276]: @pd0wm What do you think about removing these asserts, inherited from the python code? Matrix dimensions defined in line 248.

**Review by wolterhv — COMMENTED** [review]
@pd0wm  I've finished a first pass on the implementation of the functions for this draft PR, but I'm not sure what tests you referred to in your comment (https://github.com/commaai/openpilot/issues/2386#issuecomment-815741292). Are you referring to the tests for `locationd` from the `openpilot` repo?

File: rednose/helpers/ekf_sym.h [hunk 3]
```diff
@@ -62,8 +62,8 @@ class EKFSym {
 
   void augment();
   Eigen::VectorXd get_augment_times();
-  // def rts_smooth(self, estimates, norm_quats=False):
-  Eigen::MatrixXd rts_smooth(UNKNOWN_TYPE estimates, bool norm_quats); // TODO norm_quats shall default to false
+  std::vector<std::vector<MatrixXdr>> rts_smooth(UNKNOWN_TYPE estimates, bool norm_quats = false);
```
[the remark this answers is not in the document]
Comment (wolterhv) [reply rednose/helpers/ekf_sym.h]: In the meantime I've opted for a struct which can store the two tensors, following the approach used with the `Observation` and `Estimate` structs.

**Comment by wolterhv** [comment]
The `chi2_lookup` function could be optimised further by 
1) skipping the reading of lookup table file header and metadata, in the way assuming that lookup table file is correct and that numbers are stored as doubles and 
2) just fetching the boundary values needed for the interpolation instead of the whole row.

The current python implementation relies on the `__file__` dynamic attribute to determine the location of the lookup table file. Compiled applications have naturally no knowledge of the location of their source files, so probably a fixed path will have to replace this in the C++ implementation. @pd0wm  what is your opinion on this?

**Comment by pd0wm** [comment]
There really is no need to parse the npy file. Just generate the table as plaintext and include it in a header file.

**Pull request closed without merging**